gcc/rust/resolve/rust-ast-resolve-type.h

```diff
@@ -193,24 +193,52 @@ class ResolveRelativeTypePath
       = ResolveTypeToCanonicalPath::resolve (path,
 					     canonicalize_type_with_generics,
 					     true);
-    if (canonical_path.is_error ())
+    if (canonical_path.is_empty ())
       {
 	rust_error_at (path.get_locus (),
 		       "Failed to resolve canonical path for TypePath");
 	return UNKNOWN_NODEID;
       }
 
     CanonicalPath lookup = canonical_path;
-    if (!prefix.is_error ())
+    if (!prefix.is_empty ())
       lookup = prefix.append (canonical_path);
 
     auto resolver = Resolver::get ();
     NodeId resolved_node = UNKNOWN_NODEID;
+
+    // Lookup type in the following priority order:
```

Comment:
I think here, I need to what's documented here : https://github.com/rust-lang/rust/blob/1f94abcda6884893d4723304102089198caa0839/compiler/rustc_resolve/src/lib.rs#L1722
When the type can't be found in the type NS, then we need to look for it in modules (and others element as documented). Is my understanding correct here? Currently, I'm trying to see if this would work by doing a global search in all Ribs (instead of targeting only the ones related to the module). It seems to work, but then a similar lookup is done so I need do apply the same workaround.

Reply:
This is interesting to read, I was thinking the fixup we added yesterday here:

which will do a scan toplevel with an empty path prefix a 2nd time which was causing an assertion inside insert_new_definition which was fixed with my PR earlier today should solve this.

I think I would like to see if you can revert this change to the type path name scope resolution and see if the fixes from yesterday and today still allow your test cases to pass.

Basically what happens is when the name resolver starts at the moment we endup with all the AST::Items canonical path's within the top-level scope so for example:


```rust
mod foo {
    pub struct Foo {
        pub f: i32,
    }

    pub fn foofn() -> i32 {
        return 3i32;
    }
}

struct Bar {b:i32}
```

The first scan toplevel gives us the paths:

[foo::Foo , foo:foofn, Bar]

But when we do the ResolveItems and we hit the Module we should do the resolve toplevel on each item then resolve item on each item so we end up with:

[Foo, foofn]
[foo::Foo , foo:foofn, Bar]

This means the relative paths of each item should be resolveable within the new rib that has been created.

The solution inside rustc here is kind of clever. The name resolver inside gccrs might need some thought because, one of the cases we don't funny support properly is the case of AST:::Items within blocks which can be forward declared and constantly having to do scan toplevel is getting a bit confusing.

Lets revert this change, see how it works and maintain a comment to the piece of source code from rustc which might provide a cleaner solution down the line.